# Methods

### Initial Exploratory Phase

##### Acquire the genome
Use the _Ranitomeya imitator_ genome found at: 
       
    /mnt/lustre/macmaneslab/shared/Stuckert/Stuckert_Ranitomeya_imitator_genome

##### Find Duplicates

Run BUSCO on the genome and view the [full_table_Ranitomeya_imitator_genome_BUSCO.tsv](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/full_table_Ranitomeya_imitator_genome_BUSCO.tsv) file to pull out gene duplications.

    See the script: Initial_BUSCO_run_pull_outDups.sh (Below)
   [Initial_BUSCO_run_pull_outDups.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/Initial_BUSCO_run_pull_outDups.sh) 

Pull out only the duplicated genes from the full summary table and concatinate them into a single file
    
    grep "Duplicated" full_table_Ranitomeya_imitator_genome_BUSCO.tsv > duplicated_BUSCOoutput.tsv

##### Preliminary Analysis

Open [file generated by the grep command](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/duplicated_BUSCOoutput.tsv) in Excel

Add the column headers in the file as such: 

    # These mirror the BUSCO column headers

    BUSCO ID    Status    Scaffold    Start    End    Score    Length

Delete string characters around each of the numbers in the saffold column.

    Find and replace ->  "scaffold" replace with nothing
    Find and replace ->  "_pilon_pilon" replace with nothing
    
    #This leaves behind only the numbers which I can work with in R
    #Save as a CSV, TSVs are a bit cumbersome
    
The table now looks like this: [duplicated_genes.csv](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/duplicated_genes.csv)
    
Use R to determine statistics about the dataset

    See the script: initial_R_dataAnalysisAndFigures.R (Below)
    
[initial_R_dataAnalysisAndFigures.R](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/initial_R_dataAnalysisAndFigures.R)

View the CSV generated by R with the Number of Deplicates per Contig Here: [duplicates_perContig.csv](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/duplicates_perContig.csv)

##### Try to do a test alignment with two genes found very far away from each other

Used the BUSCO ID and Scaffold from the [full_table_Ranitomeya_imitator_genome_BUSCO.tsv](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/full_table_Ranitomeya_imitator_genome_BUSCO.tsv)

Sampled a gene that was duplicated twice on scaffolds that were far apart in terms of the order they were assembled

       # In this case it was on scaffolds 9 and 6848
       # I need to unwrap the genome inorder to grep the sequence from it
       
       awk '{if(NR==1) {print $0} else {if($0 ~ /^>/) {print "\n"$0} else {printf $0}}}' imitator.1.3.6.fa > unwrapped_imitator.1.3.6.fa
       
     
________________________________________________________________________________________________________________________________________
**_This was used to align an entire scaffold which is largely impractical. Don't need to do this, can just pull out the region of interest_**

       # I pulled out each of the scafold chunks and put them into their own files
       
       grep "scaffold9_pilon_pilon" -A1 unwrapped_imitator.1.3.6.fa > scaffold9_pilon_pilon.fa
       grep "scaffold6848_pilon_pilon" -A1 unwrapped_imitator.1.3.6.fa> scaffold6848_pilon_pilon.fa
       
       # I cat them together
       
       cat scaffold6848_pilon_pilon.fa scaffold9_pilon_pilon.fa > MAFFTtrial.fa
       
Take this file and test run a MAFFT alignment

    See the script: MAFFTtrial.sh (Below)
    
[MAFFTtrial.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/MAFFTtrial.sh)

________________________________________________________________________________________________________________________________________
**_I did it again but this time with only the portion of the scaffold that was of interest_**

Need to load linuxbrew/colsa so I can use bedtools [getfasta](https://bedtools.readthedocs.io/en/latest/content/tools/getfasta.html) function

       bedtools getfasta will extract the sequence defined by the coordinates in a BED interval 
       and create a new FASTA entry in the output file for each extracted sequence. By default, 
       the FASTA header for each extracted sequence will be formatted as follows: 
       “<chrom>:<start>-<end>”.
       
Command:

       bedtools getfasta -fi unwrapped_imitator.1.3.6.fa -fo regions_of_interestTEST.fa -bed id_file.tsv
       
       -fo	Specify an output file name. By default, output goes to stdout.
       -fi    Input fasta file, i used the unwrapped genome
       -bed   ID file that is required with gene coordinates
       #For -bed I pulled out the given beginning and ending coordinates of the gene of interest

See the ID file (-bed): id_file.tsv (Below)
    
[id_file.tsv](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/id_file.tsv)

Take the bedtools output fasta file and test run a MAFFT alignment

    See the script: MAFFTtest_regions_of_interest.sh (Below)
    
[MAFFTtest_regions_of_interest.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/MAFFTtest_regions_of_interest.sh)

**View Alignment Here:** [test_alignment_of_EOG090701ET.aln](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/test_alignment_of_EOG090701ET.aln)

**View on NCBI Alignment Viewer:** [NCBI Alignment View](https://www.ncbi.nlm.nih.gov/projects/msaviewer/?key=phY8y7oUZcG5Slu-mlttuM0omiildtd53XTYXclIzlv_y84R8vfLRqqpH50JPlDLEpsJxxrYPYI3xC7RJNYh2yXoD9QG-TnS,oBA6zbwSY8e_TF24nF1rvssunC6gcNJ_2HLdW8xOy136zcsX9_HOrP7xtcX4IMZbhAufV4xIqxKhVLhBska3S7N4mUSQaa9C)
________________________________________________________________________________________________________________________________________
**_I did this a third time but this time I extended the region of interest 3kb in either direction to gather an idea for what is on either side of the gene_**

Need to load linuxbrew/colsa so I can use bedtools [getfasta](https://bedtools.readthedocs.io/en/latest/content/tools/getfasta.html) function
       
Command:

       bedtools getfasta -fi genomeFiles/unwrapped_imitator.1.3.6.fa -fo regions_of_interestTEST-3kb.fa -bed id_file-3kb.tsv
       
       See the ID file (-bed): id_file.tsv (Below)
    
[id_file-3kb.tsv](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/id_file-3kb.tsv)

Take the bedtools output fasta file and test run a MAFFT alignment

    See the script: MAFFTtest_regions_of_interest-3kb.sh (Below)
    
[MAFFTtest_regions_of_interest-3kb.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/MAFFTtest_regions_of_interest-3kb.sh)

**View Alignment Here:** [test_alignment_of_EOG090701ET-3kb.aln](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/test_alignment_of_EOG090701ET-3kb.aln)

**View on NCBI Alignment Viewer:** [NCBI Alignment View-3kb](https://www.ncbi.nlm.nih.gov/projects/msaviewer/?key=1WVPuMlnFrLKOSjN6Sgey75b6xAHTnVBf0x6ZWtwbGNd82wr1a7sSwypRp1GYJR81izNcN5v-TXzc-pm4GHlbOFfy2PCTv1l,xHReqdh2B6PbKDnc-DkP2q9K-gEIX3pQcF11dGRhY3JS4mM62r_jXr0M-DjKMY10zyTUeMdn4D3qe_Nu-Wn8ZPhX0mvbRuRt)

________________________________________________________________________________________________________________________________________
________________________________________________________________________________________________________________________________________

### Analyze the depth of the sequencing:

#### Analyze the depth of the sequencing for the Nanopore reads

Run a BWA alignment and use samtools depth to check the depth

    See the script: NanoporeBWAandDepth.sh (Below)
    
[NanoporeBWAandDepth.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/NanoporeBWAandDepth.sh)

#### Analyze the depth of the sequencing for the Illumina reads

Run a BWA alignment and use samtools depth to check the depth

    See the script: IlluminaBWAandDepth.sh (Below)
    
[IlluminaBWAandDepth.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/IlluminaBWAandDepth.sh)

#### Analyze the depth of the sequencing for the PacBio reads

Run a BWA alignment and use samtools depth to check the depth

    See the script: PacBioBWAandDepth.sh (Below)
    
[PacBioBWAandDepth.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/PacBioBWAandDepth.sh)

### Look for over representation in reads (example: Nanopore)

       #Reduce the file to just be the contigs with a depth larger than 8
       awk -v x=8 '$3 >= x'  nanoDepth.tsv > nanoDepthgreaterthan8.tsv
       
       #Sort Reduced File
       sort -k 3,3  nanoDepthgreaterthan8.tsv > nanoDepthgreaterthan8SORTED.tsv
       
### Filter Nanopore data to see if it reduces the crazy over representation

Run samtools view command to filter

       See the script: NanoporeFilter-f4.sh (Below)
      
[NanoporeFilter-f4.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/NanoporeFilter-f4.sh)

Do it again with different flag and compare filtering ability

       See the script: NanoporeFilter-f256.sh (Below)
      
[NanoporeFilter-f256.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/NanoporeFilter-f256.sh)

representation
      
[IlluminaFilter-f4.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/IlluminaFilter-f4.sh)

Do it again with different flag and compare filtering ability

       See the script: IlluminaFilter-f256.sh (Below)
      
[IlluminaFilter-f256.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/IlluminaFilter-f256.sh)

### Run statistics on the Nanopore data of the new filtered data

Run samtools flagstat command

       See the script: NanoporeFlagstat-f4.sh (Below)
      
[NanoporeFlagstat-f4.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/NanoporeFlagstat-f4.sh)

Do it again with different the data from the 256 flag

       See the script: NanoporeFlagstat-f256.sh (Below)
      
[NanoporeFlagstat-f256.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/NanoporeFlagstat-f256.sh)

### Run statistics on the Illumina data of the new filtered data

Run samtools flagstat command

       See the script: IlluminaFlagstat-f4.sh (Below)
      
[IlluminaFlagstat-f4.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/IlluminaFlagstat-f4.sh)

Do it again with different the data from the 256 flag

       See the script: IlluminaFlagstat-f256.sh (Below)
      
[IlluminaFlagstat-f256.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/IlluminaFlagstat-f256.sh)

### Run Depth on the filtered sam files (Nanopore)

Run samtools depth command on both files

       See the script: Nanopore_postfilterDepth.sh (Below)
      
[Nanopore_postfilterDepth.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/Nanopore_postfilterDepth.sh)


### Run Depth on the filtered sam files (Illumina)

Run samtools depth command on both files

       See the script: Illumina_postfilterDepth.sh (Below)
      
[Illumina_postfilterDepth.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/Illumina_postfilterDepth.sh)

### I found out the filters didnt seem to do much. So I just proceeded with the regular depth TSV files
#### I then ran some stats to see what the files look like

This is the script I ran for Illumna, I did the Nanopore with the same procedure though it was on the head node instead in a tmux window

       See the script: IlluminaDepthStats.sh (Below)
      
[IlluminaDepthStats.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/IlluminaDepthStats.sh)

I ran another command in a tmux window to sort the file to get the min and the max depth

       sort -k 3,3  IlluminaDepth.tsv > illumina_readsDepthSORTED.tsv

### I wrote a script to calculate summary stats for the depth file

I initially did it in bash which worked but for larger files it ran out of memory and it took a long time to run.

       awk '{sum+=$3;a[x++]=$3;b[$3]++}b[$3]>Mode{Mode=$3}END{print "Mean: " sum/x "\nMedian: "a[int((x-1)/2)]"\nMode: " Mode}' nanopore_readsDepth.tsv > median_output_nanopore_readsDepth.tsv
       
So I wrote a Python script to do it in a more efficient way and submitted it in a shell script

See the script: IlluminaDepthStats.sh (Below)
      
[IlluminaDepthStats.sh](https://github.com/TroyLaPolice/Gene-Duplication-Origin-in-a-Non-Model-Organism/blob/master/IlluminaDepthStats.sh)
       
       
